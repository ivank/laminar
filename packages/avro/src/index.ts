/**
 * @packageDocumentation
 * @module @laminar/avro
 */
export { AvroDecimal, bufferToDecimal, decimalToBuffer, to64Bit } from './avro-decimal';
export { AvroTimestampMillis } from './avro-timestamp-millis';
export { AvroEpochDays } from './avro-epoch-days';
