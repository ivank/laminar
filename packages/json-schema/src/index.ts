/**
 * @packageDocumentation
 * @module @laminar/json-schema
 */

export { ValidationError } from './ValidationError';
export { ResolveError } from './ResolveError';
export {
  InvalidCode,
  Invalid,
  Messages,
  Options,
  Validator,
  Validation,
  error,
  empty,
  onlyErrors,
  errors,
  evaluateItem,
  evaluateProperty,
  skipRest,
  hasErrors,
  hasEvaluatedItem,
  hasEvaluatedProperty,
  combine,
  combineErrors,
  childOptions,
  validateSchema,
} from './validation';
export { Schema, JsonSchema, PrimitiveType } from './schema';
export { toMessages, toMessage, messages } from './messages';
export {
  isCompiled,
  withinContext,
  toSchemaObject,
  compile,
  ResolvedSchema,
  resolve,
  resolveFile,
  RefMap,
  FileContext,
} from './resolve';
export {
  validate,
  validateCompiled,
  ensureValid,
  ensureValidCompiled,
  ValidateOptions,
  ValidateSchemaOptions,
  ValidateCompiledOptions,
  Result,
  ResultError,
  ResultSuccess,
  Drafts,
  FormatErrors,
} from './validate';
export { CoerceCompiledOptions, CoerceSchemaOptions, CoerceOptions, coerce, coerceCompiled } from './coerce';
export { coerceSchema, coercers, CoercerOptions, Coercer } from './coercion';
