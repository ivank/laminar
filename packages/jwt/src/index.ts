/**
 * @packageDocumentation
 * @module @laminar/jwt
 */

export { RequestAuthInfo, RequestSign, User, Session, JWTData, JWTSign, JWTVerify, VerifyJWTData } from './types';
export { jwkPublicKey, JWKHandlerOptions } from './jwk';
export {
  authMiddleware,
  createSession,
  verifyBearer,
  jwtSecurityResolver,
  verifyToken,
  createSessionMiddleware,
  verifyJWT,
} from './jwt';
export {
  verifyKeycloack,
  keycloakAuthMiddleware,
  keycloakJwtSecurityResolver,
  JWTDataKeycloak,
  KeycloakOptions,
} from './keycloak';
