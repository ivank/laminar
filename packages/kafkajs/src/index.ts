/**
 * @packageDocumentation
 * @module @laminarjs/kafkajs
 */
export { KafkaConsumerService } from './kafka-consumer.service';
export { KafkaConsumerOptionalService } from './kafka-consumer-optional.service';
export {
  KafkaProducerService,
  RegisterSchemas,
  producerMiddleware,
  ProducerContext,
  registerSchemas,
  toProducerRecord,
  ToProducerRecordConfig,
  RegisterSchemasConfig,
} from './kafka-producer.service';
export { kafkaLogCreator, produce, Produce } from './helpers';
export {
  DecodedKafkaMessage,
  DecodedEachMessagePayload,
  DecodedBatch,
  DecodedEachBatchPayload,
  EachBatchConsumer,
  EachMessageConsumer,
  EncodedMessage,
  EncodedProducerRecord,
  SchemaRegistryConsumerRunConfig,
  EncodedSchemaProducerRecord,
  EachMessageeMiddleware,
  EachBatchMiddleware,
} from './types';
export { KafkaAdminService, CreateTopics } from './kafka-admin.service';
