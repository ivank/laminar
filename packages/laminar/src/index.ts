/**
 * @packageDocumentation
 * @module @laminarjs/laminar
 */
export {
  toIncomingMessageResolver,
  toRequestListener,
  HttpService,
  HttpServiceParams,
  HttpsServiceParams,
  IncomingMessageResolverParams,
  HttpParams,
} from './http/http-service';
export {
  optional,
  redirect,
  file,
  response,
  ok,
  noContent,
  movedPermanently,
  found,
  seeOther,
  notModified,
  badRequest,
  unauthorized,
  forbidden,
  notFound,
  internalServerError,
  json,
  yaml,
  form,
  binary,
  pdf,
  xml,
  text,
  html,
  css,
  csv,
  jsonOk,
  jsonCreated,
  jsonNoContent,
  jsonMovedPermanently,
  jsonFound,
  jsonSeeOther,
  jsonBadRequest,
  jsonUnauthorized,
  jsonForbidden,
  jsonNotFound,
  jsonInternalServerError,
  textOk,
  textCreated,
  textMovedPermanently,
  textFound,
  textSeeOther,
  textBadRequest,
  textUnauthorized,
  textForbidden,
  textNotFound,
  textInternalServerError,
  htmlOk,
  htmlCreated,
  htmlMovedPermanently,
  htmlFound,
  htmlSeeOther,
  htmlBadRequest,
  htmlUnauthorized,
  htmlForbidden,
  htmlNotFound,
  htmlInternalServerError,
  yamlOk,
  yamlCreated,
  yamlMovedPermanently,
  yamlFound,
  yamlSeeOther,
  yamlBadRequest,
  yamlUnauthorized,
  yamlForbidden,
  yamlNotFound,
  yamlInternalServerError,
  setCookie,
  SetCookie,
  ResponseCreator,
  FileOptions,
} from './http/response';
export { corsMiddleware, CorsConfig, OriginChecker } from './http/middleware/cors.middleware';
export {
  BodyParser,
  parseJson,
  parseForm,
  parseText,
  parseDefault,
  defaultBodyParsers,
  bodyParserMiddleware,
  concatStream,
  parseBody,
} from './http/middleware/body-parser.middleware';
export { toHttpRequest } from './http/request';
export { parseCookie, serializeCookie, CookieOptions } from './http/cookie';
export { parseQueryObjects, toJson, Json } from './helpers';
export {
  responseTimeMiddleware,
  ResponseTimeConfig,
  defaultResponseTimeHeader,
} from './http/middleware/response-time.middleware';
export { requestLoggingMiddleware, RequestLoggingMiddlewareParams } from './http/middleware/request-logging.middleware';

export { LoggerMetadata, LoggerLike, withStaticMetadata, LoggerContext, loggerMiddleware } from './logger/index';

export { start, stop, Application, init, run, InitOrder, StartLogger, StopLogger } from './application';
export {
  responseParserMiddleware,
  defaultResponseParsers,
  jsonResponseParser,
  formResponseParser,
  ResponseParser,
  parseResponse,
} from './http/middleware/response-parser.middleware';
export {
  errorsMiddleware,
  defaultErrorHandler,
  RequestError,
  HttpErrorHandler,
} from './http/middleware/errors.middleware';
export {
  toMultipartBoundary,
  toMultipartData,
  parseMultipartBuffer,
  MultipartData,
  MultipartItem,
  MultipartParser,
} from './http/multipart-parser';
export {
  HttpContext,
  HttpResponseBody,
  HttpResponse,
  HttpListener,
  HttpMiddleware,
  IncomingMessageResolver,
} from './http/types';
export { HttpError, isHttpError } from './http/http-error';
export {
  router,
  get,
  post,
  patch,
  del,
  options,
  put,
  route,
  staticAssets,
  RouteContext,
  AppRoute,
  PathRoute,
  PathRouteOptions,
  StaticAssetsOptions,
  Matcher,
} from './http/router';
export { Empty, Service, AbstractMiddleware, Middleware } from './types';
export {
  openApi,
  securityOk,
  securityError,
  securityRedirect,
  isSecurityOk,
  defaultError,
  OapiContext,
  AppRouteOapi,
  OapiPath,
  OapiAuthInfo,
  SecurityOk,
  Security,
  RequestSecurityResolver,
  OapiSecurityResolver,
  OapiSecurity,
  ResponseOapi,
  OapiPaths,
  OapiConfig,
} from './http/open-api';

export { toErrorMetadata, LaminarError } from './errors';
export { passThroughMiddleware } from './middleware/pass-through.middleware';
