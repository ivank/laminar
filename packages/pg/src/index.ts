/**
 * @packageDocumentation
 * @module @laminar/pg
 */
export { PgContext, pgMiddleware, PgMiddlewareConfig } from './pg.middleware';
export { PgService } from './pg.service';
export { PgClient, PgClientConfig } from './pg.client';
