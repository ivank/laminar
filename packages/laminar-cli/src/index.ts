import { ResolvedSchema } from '@ovotech/json-schema';
import { printDocument } from '@ovotech/ts-compose';
import { oas31 } from 'openapi3-ts';
import { convertOapi as apiConvert } from './commands/api/convert';
import { convertOapi as axiosConvert } from './commands/axios/convert';
import { convertSchema } from './convert-schema';

export { convertSchema } from './convert-schema';

export const apiContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(
    apiConvert(
      { root: schema as oas31.OpenAPIObject & { [index: string]: unknown }, refs },
      schema as oas31.OpenAPIObject & { [index: string]: unknown },
    ),
  );

export const axiosContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(
    axiosConvert(
      { root: schema as oas31.OpenAPIObject & { [index: string]: unknown }, refs },
      schema as oas31.OpenAPIObject & { [index: string]: unknown },
    ),
  );

export const schemaContent = ({ schema, refs }: ResolvedSchema): string =>
  printDocument(convertSchema({ root: schema as oas31.SchemaObject, refs }, schema));

export { toTypeScript as avroConvert } from './commands/avro/convert';
export { toTypeScript as fixturesConvert } from './commands/fixtures/convert';
export { toExternalContext as avroExternalConvert } from './commands/avro/convert';
