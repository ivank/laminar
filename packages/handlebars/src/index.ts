/**
 * @packageDocumentation
 * @module @laminar/handlebars
 */

import { Middleware } from '@laminar/laminar';
import { readdirSync, readFileSync, existsSync } from 'fs';
import { compile, RuntimeOptions, TemplateDelegate } from 'handlebars';
import { join, relative } from 'path';
import { statSync } from 'fs';

export type Cache = Map<string, { template: TemplateDelegate; mtime: Date }>;

/**
 * Cache types.
 * - preload: load all the templates ones into memory
 * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
 * - none: do not cache templates and load them on every request.
 */
export type CacheType = 'preload' | 'expiry' | 'none';

export interface TemplateConfig {
  /**
   * Helper functions passed to handlebar templates
   */
  helpers?: RuntimeOptions['helpers'];
  /**
   * Main directory where handlebar files should be located, defaults to current working directory.
   * Looks for "partials" and "views" directories inside of it.
   */
  dir?: string;
  /**
   * Directory where all the partials are located, inside the main 'dir' directory, defaults to 'partials'
   */
  partials?: string;
  /**
   * Name of the handlebars file extension. Defaults to 'handlebars', so that only 'my-file.handlebars' kind of files are loaded.
   */
  extension?: string;
  /**
   * Directory name where all the views are located, inside the main 'dir' directory, defaults to 'views'
   */
  views?: string;
  /**
   * Cache types.
   * - preload: load all the templates ones into memory
   * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
   * - none: do not cache templates and load them on every request.
   *
   * Defaults to 'preload'
   */
  cacheType?: CacheType;
  /**
   * Global data to be passed to every template
   */
  globalData?: Record<string, unknown>;
}

export interface CompileTemplatesOptions {
  dir: string;
  extension: string;
}

export type HandlebarsRender = (view: string, data?: Record<string, unknown>) => string;

export interface HandlebarsContext {
  hbs: HandlebarsRender;
}

const hasExtension =
  (extension: string) =>
  (file: string): boolean =>
    file.endsWith(extension);

export const deepReaddirSync = ({ dir, extension }: CompileTemplatesOptions): string[] =>
  readdirSync(dir, { withFileTypes: true })
    .reduce<string[]>((all, item) => {
      return all.concat(
        item.isDirectory() ? deepReaddirSync({ dir: join(dir, item.name), extension }) : join(dir, item.name),
      );
    }, [])
    .filter(hasExtension(extension));

const toName = (dir: string, file: string): string => relative(dir, file.slice(0, file.lastIndexOf('.')));

export const compileTemplates = ({ dir, extension }: CompileTemplatesOptions): Array<[string, TemplateDelegate]> =>
  deepReaddirSync({ dir, extension }).reduce<Array<[string, TemplateDelegate]>>(
    (all, file) => [...all, [toName(dir, file), compile(readFileSync(file, 'utf8'))]],
    [],
  );

export const loadTemplate = (
  name: string,
  cache: Cache,
  cacheType: CacheType,
  { dir, extension }: CompileTemplatesOptions,
): TemplateDelegate => {
  const file = `${join(dir, name)}.${extension}`;

  const stat = cacheType === 'expiry' ? statSync(file) : { mtime: new Date(0) };
  const cached = cache.get(name);

  if (cacheType !== 'none' && cached && stat.mtime < cached.mtime) {
    return cached.template;
  } else {
    const template = compile(readFileSync(file, 'utf8'));
    cache.set(name, { template, mtime: cacheType === 'expiry' ? stat.mtime : new Date() });
    return template;
  }
};

const loadPartials = (options: CompileTemplatesOptions): { [key: string]: TemplateDelegate } =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce((acc, [name, template]) => ({ ...acc, [name]: template }), {})
    : {};

const loadTemplates = (cache: Cache, options: CompileTemplatesOptions): Cache =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce(
        (cache, [name, template]) => cache.set(name, { template, mtime: new Date() }),
        cache,
      )
    : cache;

export const handlebars = ({
  helpers = {},
  dir = process.cwd(),
  extension = 'handlebars',
  partials = 'partials',
  views = 'views',
  cacheType = 'preload',
  globalData = {},
}: TemplateConfig): HandlebarsRender => {
  const cache = cacheType === 'preload' ? loadTemplates(new Map(), { dir, extension }) : new Map();
  const partialTemplates = cacheType === 'preload' ? loadPartials({ dir: join(dir, partials), extension }) : undefined;

  return (name: string, data: Record<string, unknown> = {}) => {
    const template = loadTemplate(name, cache, cacheType, { dir: join(dir, views), extension });
    return template(
      { ...globalData, ...data },
      { helpers, partials: partialTemplates ?? loadPartials({ dir: join(dir, partials), extension }) },
    );
  };
};

export const handlebarsMiddleware = (config: TemplateConfig): Middleware<HandlebarsContext> => {
  const hbs = handlebars(config);
  return (next) => (req) => next({ ...req, hbs });
};
