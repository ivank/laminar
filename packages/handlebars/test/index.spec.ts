import { join } from 'path';
import { compileTemplates, deepReaddirSync } from '../src';

const dir = join(__dirname, 'root');

describe('Laminar Handlebars', () => {
  it('Test deepReaddirSync', () => {
    const partials = deepReaddirSync({ dir: join(dir, 'partials'), extension: 'handlebars' });

    expect(partials).toEqual([
      join(dir, 'partials/inner/condition.handlebars'),
      join(dir, 'partials/inner/deep/unless.handlebars'),
      join(dir, 'partials/layout.handlebars'),
      join(dir, 'partials/list.handlebars'),
      join(dir, 'partials/span.handlebars'),
    ]);
  });

  it('Test compileTemplates', () => {
    const partials = compileTemplates({ dir: join(dir, 'partials'), extension: 'handlebars' });

    expect(partials.find(([name]) => name === 'list')?.[1]({ list: ['one', 'two'] })).toMatchSnapshot();
    expect(partials.find(([name]) => name === 'span')?.[1]({ text: 'example' })).toMatchSnapshot();
    expect(partials.find(([name]) => name === 'inner/condition')?.[1]({ test: true })).toMatchSnapshot();
    expect(partials.find(([name]) => name === 'inner/deep/unless')?.[1]({})).toMatchSnapshot();
  });
});
