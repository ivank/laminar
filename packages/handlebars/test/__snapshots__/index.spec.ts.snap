// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Laminar Handlebars Test compileTemplates 1`] = `
"<ul>
  <li>0: one</li>
  <li>1: two</li>
</ul>
"
`;

exports[`Laminar Handlebars Test compileTemplates 2`] = `
"<span>example</span>
"
`;

exports[`Laminar Handlebars Test compileTemplates 3`] = `
"<span>has test true</span>
"
`;

exports[`Laminar Handlebars Test compileTemplates 4`] = `
"<div class="entry">WARNING</div>
"
`;
